### ISSUE TITLE
XR support?

### ISSUE THREAD

**Issue by Bossieh** [pmndrs/postprocessing#452]
Currently it seems XR is not supported, see related issue here: https://github.com/pmndrs/react-xr/issues/205

Are there plans to support XR in the future?

**Comment by vanruesc** [comment]
Right, `postprocessing` effects are currently not stereo-aware. XR support will be addressed when v7 is complete.

**Comment by CodyJasonBennett** [comment]
three.js core has long needed fixes since implementing layers. See https://github.com/mrdoob/three.js/pull/26160 which does so and implements WebXR rendering for postprocessing and three.js.

**Comment by danrossi** [comment]
> three.js core has long needed fixes since implementing layers. See [mrdoob/three.js#26160](https://github.com/mrdoob/three.js/pull/26160) which does so and implements WebXR rendering for postprocessing and three.js.

I've checked this out and merged changes and it works. Will those changes work with this project or do I continue using the EffectComposer change ? I am trying a lens distortion project and its applying to the whole viewport for both cameras, which then gets overlayed. Not sure how to correct it. 

https://github.com/ycw/three-lens-distortion

**Comment by danrossi** [comment]
I just check out this project. The demos are all embedded into a custom application with no raw example how to use. 

This example code doesn't work with the three xr fixes when going into xr. It's black rendering. How to make a custom effect on each camera rather than the output buffer also ? I assume EffectPass is applied to the whole viewport correct ?

Just adding any effect.

```
composer = new EffectComposer(renderer, {
	frameBufferType: HalfFloatType
});

 const sepiaEffect = new SepiaEffect({ blendFunction: BlendFunction.SKIP });
sepiaEffect.blendMode.opacity.value = .5;

const pass = new EffectPass(camera,
    sepiaEffect
);
composer.addPass(new RenderPass(scene, camera));
composer.addPass(pass);

renderer.setAnimationLoop(update);

function update(deltaTime) {
   composer.render(deltaTime);
}
```

**Comment by danrossi** [comment]
I've created a project example for a ported lens distortion. With/without the webxr three fix I get distortion for both cameras no issue with black rendering. But there is an issue showing the texture for some reason in xr. 

https://danrossi.github.io/three-lens-distortion/example/postprocessing.html

xr fix. With three's EffectComposer it shows up wrong. With postprocessing the distortion in both camera is showing how I expect it. Just no texture yet.

https://danrossi.github.io/three-lens-distortion/example/postprocessing-xr.html

https://github.com/danrossi/three-lens-distortion/blob/main/src/glsl/lens-distortion.frag

**Comment by danrossi** [comment]
I've updated my examples and a non video texture one shows up  with no rendering and black background also with the three xr fix sadly. 

https://danrossi.github.io/three-lens-distortion/example/postprocessing.html
https://danrossi.github.io/three-lens-distortion/example/postprocessing-video.html

**Comment by danrossi** [comment]
I've taken to task to migrate xr changes over to postprocessing with updated examples. 

https://danrossi.github.io/three-lens-distortion/example/postprocessing.html
https://danrossi.github.io/three-lens-distortion/example/postprocessing-video.html

Using these branches

https://github.com/CodyJasonBennett/three.js/tree/fix/offscreen-webxr
https://github.com/danrossi/postprocessing/blob/main/src/passes/EffectPass.js#L578

It still not showing right in the webxr polyfill with overlayed cameras. But in mixed reality simulator it shows up correctly. not sure if the effect is being applied to each xr camera or the whole viewport.

**Comment by CodyJasonBennett** [comment]
The effect is applied once to the whole viewport; the render pass should contain XR transforms. You can read from pose directly if you want to apply per-eye effects to keep it to 1 draw call -- it's notable that WebXR spec allows for any number of views which some devices have 6 or 100 of. Setting `renderer.xr.enabled = false` will opt out of XR transforms when rendering which all effects should be doing. I've already applied the necessary changes in https://codesandbox.io/s/kgjxfb which is waiting on https://github.com/mrdoob/three.js/pull/26160.

**Comment by danrossi** [comment]
Yes that works too. I'm sorry I don't understand  how to modify the effect I made to do that so the distortion is applied to both cameras seperately. I've integrated your three changes into the tests. And updated RenderPass. I've only updated EffectPass. 

https://github.com/danrossi/postprocessing/blob/main/src/passes/RenderPass.js#L322

Is those changes still needed ?

**Comment by danrossi** [comment]
I've rebuilt everything and integrated webxr into the manual demos in this branch. But npm isn't liking checking out your branch of three with the fixes, the module has to be copied in. There is a possible bug with webxrmanager I have to figure out. I'll work out the issue with my effect so it's applied to both webxr cameras. 

https://github.com/danrossi/postprocessing/tree/webxr-render-fix

**Comment by danrossi** [comment]
I've fully upgraded the manual and public demos with the fix merged with latest three in that branch. It seems to be working but can't get pages generated for it. fullscreen render.

![Screenshot (91)](https://github.com/pmndrs/postprocessing/assets/710277/5e65e692-6b62-419a-bee6-ec2b056ce622)

**Comment by Utopiah** [comment]
Wondering what's the status.

**Comment by danrossi** [comment]
I spent too much time on it and gave up for now. It's waiting on integrating changes in three. I couldnt get the lens distortion showing up stereo its just monoscopic.
